3.0 biased winner strategy

mneedham opened the pull request:
changelog: Adds configuration settings for ordering branch handling and store copying on slaves

Commit:
Get rid of the biased winner strategy

When a node is being 'demoted' we'll just kick
off a normal election rather than ensuring that
it doesn't win.

This will deal with the scenario where a master goes
away but has returned by the time vote requests are
sent out.

Before this commit we'd reject that vote request and
make a sub optimal choice of new master

Commit:
Make the choice of SwitchToSlave approach configurable

This commit makes it possible to choose which strategy is used
when branching the store.

The current approach is:
branch store -> copy down new store -> put new store in place

The alternative approach is:
copy down new store -> branch store -> put new store in place

MishaDemianenko committed:
Specific hanling of upgrade transaction during update pooling

During store upgrade in case if transaction logs are missing upgrade transaction checksum will be set to predefined constant value
(since we can't get find real value in the logs).
In HA setup upgraded store will be copied by slaves and update pullers will try to fetch updates starting from upgrade transaction id.
Master should be able to recognise that specific upgrade transaction id and use checksum that was assign to it during upgrade.

MishaDemianenko committed:
Remove redundant imports.

MishaDemianenko committed:
Update index writer delete flush settings

Add setting for max buffered delete terms, max buffered docs;
Introduce couple of additional feature toggle for already defined configuration parameters.

benbc committed:
Include neo4j-admin manpage in the Debian package

benbc committed:
Fix neo4j-admin manpage to refer to built-in help.

benbc committed:
Use static manpages rather than generating them from asciidoc

benbc committed:
Remove remaining SNIPPET tags

All the ones that are still in use have been moved to the manual
subproject. Those removed here are all obsolete.

benbc committed:
Remove some out-of-date repo documentation

benbc committed:
Move one remaining static docs file under manual

benbc committed:
Move config docs generation code into docs tools module

benbc committed:
Move doctools into manual project

benbc committed:
Remove remaining docs-related stuff from server tests

pontusmelke committed:
Handle java lists in private type conversion

pontusmelke committed:
Handle java maps in private type conversion

benbc committed:
Remove manual

benbc committed:
Remove errant docs/examples Maven references

pontusmelke committed:
Plan description fails for NestedExpression

digitalstain reviewed (commented):
We should ensure that checkstyle rules for 3.1 are met

pontusmelke committed:
Don't throw NullPointerException in LabelsFunction

If the node is deleted at the same time as we are trying to access the labels
of that node there is a significant risk that we will have a `NullPointerException`.
Instead of an error we should just return an empty iterator for that case.

Fixes #8239

digitalstain committed:
Removes unused code in cluster elections

After recent changes, the concept of promoting a cluster member
 is no longer needed and the idea of demotion has been significantly
 simplified. This commit removes the unused code coming from these
 changes.

digitalstain committed:
ElectionCredentials now extends Comparable<ElectionCredentials>

Before the type parameter on the Comparable interface was left
 to Object and lots of uses were on Comparable<Object>
 unnecessarily, causing warnings and ugly casts. This change
 simplifies things a bit and provides more type safety.

digitalstain committed:
Merge branch '3.0' into 3.0-biased-winner-strategy

digitalstain committed:
Improves Description for branched_data_copying_strategy

Mats-SX commented:
@digitalstain Do you know what Pontus' commits are doing in this PR?
